aten/src/ATen/native/cuda/TensorShape.cu

```diff
@@ -0,0 +1,399 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/Dispatch.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/ops/empty.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/util/Logging.h>
+
+namespace at::native {
+
+namespace detail {
+
+// NOTE [CUDA fast path for split_with_sizes_copy.out]
+// split_with_sizes_copy.out for contiguous operands has the following
+// properties:
+// - Each src split consists of multiple chunks that are separated by a fixed
+// stride. The number of chunks and the strides are the same across all src
+// splits.
+// - Each dst split is the concatenation of the chunks in its corresponding src
+// splits.
+// - The sizes chunks vary across splits.
+// - The alignment varies across chunks. A (src, dst) chunk pair is not
+// guaranteed to have the same alignment.
+//
+// The following strategies are employed to optimize for this workload:
+// - The entire workload is fused into a single kernel to maximize I/O
+// throughput and minimize wave quantization.
+// - To account for both small and large chunk sizes, a "jagged grid" is used.
+// Each chunk is processed by one or more blocks depending on its size.
+// - Within each chunk, the region in which writes can be vectorized is
+// identified. Within this region, writes are always vectorized and reads are
+// oppurtunistically vectorized.
+static constexpr int64_t BLOCK_SIZE = 128;
+static constexpr int64_t BYTES_PER_THREAD = 16;
+
+static __host__ __device__ inline int64_t div_up(int64_t a, int64_t b) {
+  return (a + b - 1) / b;
+}
+
+template <typename T>
+__device__ inline void stream_load128(uint4& val, const T* addr) {
+  uint64_t low, high;
```

Comment:
IIUC, `low, high` are in registers, and `val` is in registers (since defined in the device function `copy_chunk`).

I cannot find much info on the cost of register-to-register copy. I wonder if it is possible for the CUDA fast path to directly load to `reinterpret_cast<uint64_t*>(&val)[0]` and `reinterpret_cast<uint64_t*>(&val)[1]` rather than using `low, high` as intermediate. Or, is register-to-register copy so fast that there is no need?

Reply:
> I cannot find much info on the cost of register-to-register copy. I wonder if it is possible for the CUDA fast path to directly load to reinterpret_cast<uint64_t*>(&val)[0] and reinterpret_cast<uint64_t*>(&val)[1] rather than using low, high as intermediate. Or, is register-to-register copy so fast that there is no need?

It's very fast (a few clock cycles IIRC) but we are not really doing redundant moves. The compiler will (most likely) inlines this function and try to remove unnecessary intermediates, and it's really good at it.

You can put the code in compiler explorer to get a sense of what the compiler does for a call like this: https://godbolt.org/z/YdE1Trbzc

Reply:
This is very cool! Thanks for sharing the compiled snippet.